### PR TITLE
http: Adding flexible remove function to HeaderMap

### PR DESCRIPTION
**Pull request by alyssawilk**
It's fairly common in industry for companies to have their own internal headers with common prefixes (x-lyft- akamai-x-, etc) and seems generally useful to be able to clean headers going back to the public internet in a more performant way than collecting all keys with iterate() and following up with bunch of separate O(n) remove() calls. 

*Risk Level*: Low (adding unused-upstream header utility)
*Testing*: unit tests
*Docs Changes*: n/a
*Release Notes*: n/a

### CODE REVIEW

File: source/common/http/header_map_impl.cc [hunk 1]
```diff
@@ -464,6 +464,17 @@ void HeaderMapImpl::remove(const LowerCaseString& key) {
   }
 }
 
+void HeaderMapImpl::remove(const std::regex& regex) {
+  for (auto i = headers_.begin(); i != headers_.end();) {
+    absl::string_view key = i->key().getStringView();
+    if (std::regex_search(key.begin(), key.end(), regex)) {
```
Comment (jmarantz) [inline source/common/http/header_map_impl.cc]: consider std::list::remove_if ?

File: source/common/http/header_map_impl.cc [hunk 2]
```diff
@@ -464,6 +464,17 @@ void HeaderMapImpl::remove(const LowerCaseString& key) {
...
+      i = headers_.erase(i);
```
Comment (mattklein123) [inline source/common/http/header_map_impl.cc]: I think this is not going to work correctly if this is a match of an inline header. You are probably going to need to do some extra work to make it work... Can you add a test case?

File: include/envoy/http/header_map.h [hunk 3]
```diff
@@ -465,6 +466,14 @@ class HeaderMap {
    */
   virtual void remove(const LowerCaseString& key) PURE;
 
+  /**
+   * Remove all instances of headers with key matching the supplied regex.
+   * @param key_matcher supplies the regex to match header keys against. Note
+   * that this will be matching against LowerCaseString, so should be using
+   * lowercase characters.
+   */
+  virtual void remove(const std::regex& key_matcher) PURE;
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: Should we consider doing a prefix version? I suspect the regex version is going to be slower, even if heavily tuned and basically a prefix?

Comment (alyssawilk) [reply include/envoy/http/header_map.h]: Fine by me.  I'd be inclined to then swap regex for prefix as I have 3 confirmed users of prefixes and 0 that need generic regex :-)

File: source/common/http/header_map_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/header_map_impl.cc]: Hah - just converted over and got Matt's comment about inline headers.  I think the additional complexity to remove the constant time references may make this version preferable.  Will look though!

File: include/envoy/http/header_map.h (hunk 3 above)
Comment (mattklein123) [reply include/envoy/http/header_map.h]: +1

File: source/common/http/header_map_impl.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/header_map_impl.cc]: And my intuition of "I should wait an hour and see what Matt thinks" pans out once again :-D

File: include/envoy/http/header_map.h (hunk 3 above)
Comment (alyssawilk) [reply include/envoy/http/header_map.h]: got inline fixed - one more push and I'll do this swap as well

**Pull request merged**